pkg/dataobj/index/index_builder.go

```diff
@@ -0,0 +1,510 @@
+package index
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/flagext"
+	"github.com/grafana/dskit/multierror"
+	"github.com/grafana/dskit/services"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/objstore"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/bits-and-blooms/bloom/v3"
+
+	"github.com/grafana/loki/v3/pkg/dataobj"
+	"github.com/grafana/loki/v3/pkg/dataobj/index/indexobj"
+	"github.com/grafana/loki/v3/pkg/dataobj/metastore"
+	"github.com/grafana/loki/v3/pkg/dataobj/sections/logs"
+	"github.com/grafana/loki/v3/pkg/dataobj/sections/streams"
+	"github.com/grafana/loki/v3/pkg/kafka"
+	"github.com/grafana/loki/v3/pkg/kafka/client"
+)
+
+type Config struct {
+	indexobj.BuilderConfig `yaml:",inline"`
+	EventsPerIndex         int                    `yaml:"events_per_index" experimental:"true"`
+	IndexStoragePrefix     string                 `yaml:"index_storage_prefix" experimental:"true"`
+	EnabledTenantIDs       flagext.StringSliceCSV `yaml:"enabled_tenant_ids" experimental:"true"`
+}
+
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	cfg.RegisterFlagsWithPrefix("dataobj-index-builder.", f)
+}
+
+func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	cfg.BuilderConfig.RegisterFlagsWithPrefix(prefix, f)
+	f.IntVar(&cfg.EventsPerIndex, prefix+"events-per-index", 32, "Experimental: The number of events to batch before building an index")
+	f.StringVar(&cfg.IndexStoragePrefix, prefix+"storage-prefix", "index/v0/", "Experimental: A prefix to use for storing indexes in object storage. Used to separate the metastore & index files during initial testing.")
+	f.Var(&cfg.EnabledTenantIDs, prefix+"enabled-tenant-ids", "Experimental: A list of tenant IDs to enable index building for. If empty, all tenants will be enabled.")
+}
+
+type downloadedObject struct {
+	event       metastore.ObjectWrittenEvent
+	objectBytes *[]byte
+	err         error
+}
+
+const indexEventTopic = "loki.metastore-events"
+const indexConsumerGroup = "metastore-event-reader"
```

Comment:
Out of curiosity: Do we have a naming convention for topics and consumer groups?

Reply:
Unfortunately, we don't. Do you have a suggestion? I'm more than happy to change this at any time